### PR TITLE
feat: view data from multi orgs

### PR DESCRIPTION
**Pull request by tristanhcole**
**Describe the Pull Request**
Group Data from multi organisations. Currently supports only the analytics and map pages. All other pages aren't supported at this stage.
Screen Shot 2021-01-29 at 5 18 00 pm

**Todo**

- [x] Link relevant [trello card](https://trello.com/b/PA2LhlOh/sempo-dev) or [related issue](https://github.com/teamsempo/SempoBlockchain/issues) to the pull request
- [x] Request review from relevant @person or team
- [x] QA your pull request. This includes running the app, login, testing changes, checking [accessibility](https://www.a11yproject.com/checklist/) etc.
- [x] Bump [backend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/config.py#L7) and/or [frontend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/app/package.json#L3) versions following Semver

### CODE REVIEW

**Comment by tristanhcole** [comment]
[(22 hrs) Improved support for multiple ERC-20 tokens](https://trello.com/c/EKfoGDxU/609-22-hrs-improved-support-for-multiple-erc-20-tokens)

**Comment by tristanhcole** [comment]
[Master Dashboard visualizing data from multiple organisations](https://trello.com/c/nfyQUHZy/466-master-dashboard-visualizing-data-from-multiple-organisations)

File: app/client/components/dashboard/MasterWalletCard.jsx [hunk 1]
```diff
@@ -31,7 +31,7 @@ class MasterWalletCard extends React.Component {
     const amountDisbursed = creditTransferStats.total_distributed / 100;
     const amountWithdrawn = creditTransferStats.total_withdrawn / 100;
     const amountReclaimed = creditTransferStats.total_reclaimed / 100;
-    const symbol = activeToken.symbol;
+    const symbol = activeToken && activeToken.symbol;
```
Comment (tristanhcole) [inline app/client/components/dashboard/MasterWalletCard.jsx:34]: We need to check `activeToken` exists across the app, as when you change org, it temporarily is undefined

File: app/client/components/navBar/index.tsx [hunk 2]
```diff
@@ -127,15 +129,24 @@ class NavBar extends React.Component<Props, State> {
                 </NavLink>
               </Menu.Item>
             </SubMenu>
-            <Menu.Item key="/accounts" icon={<TeamOutlined translate={""} />}>
+            <Menu.Item
+              key="/accounts"
+              icon={<TeamOutlined translate={""} />}
+              disabled={isMultiOrg || false}
+            >
```
Comment (tristanhcole) [inline app/client/components/navBar/index.tsx:136]: disable nav items when multiorg

File: app/client/components/navBar/page.tsx [hunk 3]
```diff
@@ -34,6 +40,16 @@ const Page: React.FunctionComponent<OuterProps> = props => {
...
 
+  const login: LoginState = useSelector((state: ReduxState) => state.login);
+  const { organisationIds } = login;
+  const multiOrgActive = organisationIds && organisationIds.length > 1;
+
+  if (multiOrgActive && !isMultiOrg) {
+    // Trying to access a page with multi org active
+    browserHistory.push("/");
+    message.error("This page is unsupported with multi organisation");
+  }
```
Comment (tristanhcole) [inline app/client/components/navBar/page.tsx:51]: protect views from being accessed when multi org is active and page isn't `isMultiOrg=true`. this protects users from going to a page directly in the URL

File: app/client/components/organizationWrapper.jsx [hunk 4]
```diff
@@ -2,25 +2,42 @@ import React, { Component } from "react";
...
+        organisations.toString() === query_params["query_organisations"]
+      ) {
+        // Multi Org Query
+        let queryOrgArray = query_params["query_organisations"].split(",");
+        let validOrgs = queryOrgArray.every(orgId =>
+          organisations.includes(orgId)
+        );
+
+        if (validOrgs) {
+          this.props.updateActiveOrgRequest(queryOrgArray);
```
Comment (tristanhcole) [inline app/client/components/organizationWrapper.jsx:40]: update wrapper to support multi org

File: app/client/reducers/auth/loginReducer.ts [hunk 5]
```diff
@@ -37,8 +39,8 @@ export const login = (state = initialLoginState, action: LoginAction) => {
       return { ...state, isLoggingIn: true };
     case LoginActionTypes.UPDATE_ACTIVE_ORG:
       return {
-        ...state,
-        organisationId: action.payload.organisationId
+        ...state
+        // organisationIds: action.payload.organisationIds
```
Comment (tristanhcole) [inline app/client/reducers/auth/loginReducer.ts:43]: we don't need to set state as page is refreshed

File: app/client/sagas/authSagas.ts [hunk 6]
```diff
@@ -251,6 +259,14 @@ function* refreshToken() {
     if (token_request.auth_token) {
       storeSessionToken(token_request.auth_token);
 
+      let orgId = getOrgIds();
+      let orgIds = orgId && orgId.split(",");
+      if (orgIds && orgIds.length > 1) {
+        token_request["organisation_ids"] = orgIds;
+      } else {
+        token_request["organisation_ids"] = null;
+      }
```
Comment (tristanhcole) [inline app/client/sagas/authSagas.ts:268]: This made the most sense to me. As `saveOrgId` stores org ids in the web browser local storage, it made sense on reauth to check if multi org is active when reauth occurs

File: app/server/api/metrics_api.py [hunk 7]
```diff
@@ -84,6 +84,7 @@ def get(self):
         return make_response(jsonify(response_object)), 200
 
 class FiltersApi(MethodView):
+    @multi_org
```
Comment (tristanhcole) [inline app/server/api/metrics_api.py:87]: Filter needs to support multi org or alternatively be disabled on the frontend. @michielderoos what do you think?

Comment (michielderoos) [reply app/server/api/metrics_api.py:87]: This is reasonable!

**Review by michielderoos — APPROVED** [review]
Tested, LGTM! I think the UX makes a lot of sense.

**Review by enjeyw — APPROVED** [review]
Love the UX solution here! Good to merge in I think.

**Pull request merged**